libraries/core_libs/consensus/src/final_chain/final_chain.cpp

```diff
@@ -316,15 +338,30 @@ class FinalChainImpl final : public FinalChain {
     blk_header.extra_data = extra_data;
     dev::BytesMap trxs_trie, receipts_trie;
     dev::RLPStream rlp_strm;
-    for (size_t i(0); i < transactions.size(); ++i) {
-      auto const& trx = transactions[i];
-      auto i_rlp = util::rlp_enc(rlp_strm, i);
+    auto receipt_idx = 0;
+    for (; receipt_idx < transactions.size(); ++receipt_idx) {
+      const auto& trx = transactions[receipt_idx];
+      auto i_rlp = util::rlp_enc(rlp_strm, receipt_idx);
       trxs_trie[i_rlp] = trx->rlp();
-      auto const& receipt = receipts[i];
+
+      const auto& receipt = receipts[receipt_idx];
       receipts_trie[i_rlp] = util::rlp_enc(rlp_strm, receipt);
       db_->insert(batch, DB::Columns::final_chain_receipt_by_trx_hash, trx->getHash(), rlp_strm.out());
-      auto bloom = receipt.bloom();
-      blk_header.log_bloom |= bloom;
+
+      blk_header.log_bloom |= receipt.bloom();
+    }
+    if (system_transactions.size() && receipts.size() == (transactions.size() + system_transactions.size())) {
```

Comment:
when can it happen that receipts.size() != (transactions.size() + system_transactions.size()) ?